Acrolinx.Sidebar/AcrolinxSidebar.cs

```diff
@@ -1,698 +1,715 @@
+            DETAIL
+        }
+        public void RegisterClientComponent(Assembly assembly, string humanReadableName, SoftwareComponentCategory category)
+        {
+            Contract.Requires(assembly != null);
+            Contract.Requires(humanReadableName != null);
+            if (assembly != null)
+            {
+                var name = assembly.GetName();
+                RegisterClientComponent(name.Name, humanReadableName, name.Version.ToString(), category);
+            }
+        }
+
+        private bool mainComponentSet = false;
+        public void RegisterClientComponent(string id, string name, string version, SoftwareComponentCategory category)
+        {
+            Contract.Requires(id != null);
+
+            mainComponentSet |= category == SoftwareComponentCategory.MAIN;
+
+            var clientComponents = InitParameters["clientComponents"].Value<JArray>();
+
+            var component = new JObject();
+            component.Add("id", id.ToLower());
+            component.Add("name", name);
+            component.Add("version", version);
+            component.Add("category", category.ToString().ToUpper());
+            if (category != SoftwareComponentCategory.DETAIL)
+            {
+                clientComponents.Add(component);
+            }
+            Logger.AcroLog.Info(name + "\t" + version);
+        }
+
+        public String Check(IDocument document)
+        {
+            acrolinxPlugin.Document = document;
+            Logger.AcroLog.Debug("Content: " + document.Content);
+
+            var code = "new function(){var c = window.external.getContent(); "
+                + "return acrolinxSidebar.checkGlobal(c, {inputFormat:'" + document.Format.ToString().ToUpper() + "', requestDescription:{documentReference: '"
+                + document.Reference.Replace("\\", "\\\\").Replace("'", "\\'").Replace("\n", "").Replace("\r", "") + "'}, selection:{ranges:" + SerializeSelection(document.Selections) + "}})}();";
+
+            dynamic check = Eval(code);
+            string checkId = check.checkId;
+            return checkId;
+        }
+        public void CancelCheck()
+        {
+            var code = "acrolinxSidebar.onGlobalCheckRejected();";
+            Eval(code);
+            Logger.AcroLog.Info("Check canceled by Acrolinx Integration.");
+        }
+
+        /// <summary>
+        /// Show a message in the Sidebar.
+        /// Supported since Acrolinx Platform 2021.2 (Sidebar version 14.28).
+        /// </summary>
+        public void ShowMessage(Util.Message.Message message)
+        {
+            var title = GetJavaScriptFriendlyString(message.Title);
+            var text = GetJavaScriptFriendlyString(message.Text);
+            var code = $"acrolinxSidebar.showMessage({{type: '{message.Type.ToString().ToLowerInvariant()}', title: '{title}', text: '{text}'}});";
+            Eval(code);
+        }
+
+        private string GetJavaScriptFriendlyString(string input)

```

Comment:
Can this be moved inside Eval()? This will remove overhead of converting string from all places e.g. check()

Reply:
We cannot put this to Eval method as we shouldn't replace the string values of the whole Eval expression. Only JavaScript parameter need to be modified, I've changed the method to "GetJavaScriptFriendlyParameterString" so it's more clear.

Reply:
Look ok to me as we are doing the string modification for Document.Reference too. Line(409)